SU2_CFD/src/numerics/CLookUpTable.cpp

```diff
@@ -0,0 +1,593 @@
+      exit_code = 1;
+    }
+
+  } else {
+    // if (rank == MASTER_NODE) cout << "WARNING: LookUp_ProgEnth: lookup is outside of table bounding box, c,h = "<<
+    // val_prog<< " "<<val_enth<<endl;
+
+    /* if point is outside of table range, search nearest neighbor */
+    nearest_neighbor = FindNearestNeighborOnHull(val_prog, val_enth, name_prog, name_enth);
+    exit_code = 1;
+  }
+
+  /* loop over variable names and interpolate / get values */
+  for (long unsigned int i_var = 0; i_var < val_names_var.size(); ++i_var) {
+    if (val_names_var.at(i_var).compare("NULL") == 0) {
+      *val_vars.at(i_var) = 0.0;
+    } else {
+      if (exit_code == 0)
+        *val_vars.at(i_var) = Interpolate(GetData(val_names_var.at(i_var)), (triangles.at(id_triangle)), interp_coeffs);
+      else
+        *val_vars.at(i_var) = GetData(val_names_var.at(i_var)).at(nearest_neighbor);
+    }
+  }
+  return exit_code;
+}
+
+void CLookUpTable::GetInterpCoeffs(su2double val_x, su2double val_y, vector<vector<su2double> >& interp_mat_inv,
+                                   vector<su2double>& interp_coeffs) {
+  vector<su2double> query_vector;
+  query_vector.push_back(1);
+  query_vector.push_back(val_x);
+  query_vector.push_back(val_y);
+
+  su2double d;
+  for (int i = 0; i < 3; i++) {
+    d = 0;
+    for (int j = 0; j < 3; j++) {
+      d = d + interp_mat_inv.at(i).at(j) * query_vector.at(j);
+    }
+    interp_coeffs.at(i) = d;
+  }
+}
+
+su2double CLookUpTable::Interpolate(const vector<su2double>& val_samples, vector<unsigned long>& val_triangle,
+                                    vector<su2double>& val_interp_coeffs) {
+  su2double result = 0;
+  su2double z = 0;
+
+  for (int i_point = 0; i_point < N_POINTS_TRIANGLE; i_point++) {
+    z = val_samples.at(val_triangle.at(i_point));
+    result += val_interp_coeffs.at(i_point) * z;
+  }
+
+  return result;
+}
+
+unsigned long CLookUpTable::FindNearestNeighborOnHull(su2double val_prog, su2double val_enth, string name_prog,
+                                                      string name_enth) {
+  su2double min_distance = 1.e99;
+  su2double next_distance = 1.e99;
+  su2double next_prog_norm;
+  su2double next_enth_norm;
+  unsigned long neighbor_id = 0;
+
+  const vector<su2double>& prog_table = GetData(name_prog);
+  const vector<su2double>& enth_table = GetData(name_enth);
+
+  su2double norm_coeff_prog = 1. / (limits_table_prog[1] - limits_table_prog[0]);
+  su2double norm_coeff_enth = 1. / (limits_table_enth[1] - limits_table_enth[0]);
+  su2double val_prog_norm = val_prog / (limits_table_prog[1] - limits_table_prog[0]);
+  su2double val_enth_norm = val_enth / (limits_table_enth[1] - limits_table_enth[0]);
+
+  for (unsigned long i_point = 0; i_point < n_hull_points; ++i_point) {
+    next_prog_norm = prog_table.at(hull.at(i_point)) * norm_coeff_prog;
+    next_enth_norm = enth_table.at(hull.at(i_point)) * norm_coeff_enth;
```

Comment:
do not use `at` on std::vector, it costs performance. su2vector gives you bounds checking in debug mode and no overhead in optimized builds

Reply:
so this:
`const vector<su2double>& prog_table = GetData(name_prog);`
becomes a su2vector,
`const su2vector<su2double>& prog_table = GetData(name_prog);`
 and also GetData will return an su2vector,
so getdata does this now:
`return table_data.at(ix_var);`

can we get an su2vector from an su2matrix table_data?

Reply:
Not an su2vector but you can get a pointer su2matrix[row]

Reply:
There is also su2matrix.innerIter(row) which also works on column-major matrices.

Reply:
I did not see a place in the code where something like that was done.
M[1] gives you a pointer to the start of the row, the number of elements is M.cols()

Reply:
Yes, so this actually works for initializing vectors from an su2matrix:

`vector<su2double> Z(X[0],X[0]+X.cols());`

It looks like we cannot initialize an su2vector from an su2matrix like that.

Reply:
Where and why do you need that? It's not going to be efficient.

Reply:
I am changing su2vector<su2vector>> table_data into an su2matrix. In CLookUpTable.hpp I now do this below for instance for GetData to work. Maybe this can be changed so it only returns a pointer? 

```
inline const vector<su2double> GetData(string name_var) {
    int ix_var = GetIndexOfVar(name_var);
 
    vector<su2double> tableDataRow(table_data[ix_var],table_data[ix_var]+table_data.cols()); 
   
    return tableDataRow;
  }
```

Reply:
That would be better, otherwise this ends up worse than vector<vector>

Reply:
I am not sure how to change this. A reinterpret_cast to create a pointer to a vector<su2double>?

Reply:
Please highlight the part of the code where this is a problem.

Reply:
It's about the code given above, the GetData function. If we keep the output the same, i.e. a vector<su2double> (or pointer to) then how to construct this from table_data without doing an element-wise copy?